files/scripts/teamd.sh

```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
+
+function debug()
+{
+    /usr/bin/logger $1
+    /bin/echo `date` "- $1" >> ${DEBUGLOG}
+}
+
+function check_warm_boot()
+{
+    SYSTEM_WARM_START=`$SONIC_DB_CLI STATE_DB hget "WARM_RESTART_ENABLE_TABLE|system" enable`
+    SERVICE_WARM_START=`$SONIC_DB_CLI STATE_DB hget "WARM_RESTART_ENABLE_TABLE|${SERVICE}" enable`
+    if [[ x"$SYSTEM_WARM_START" == x"true" ]] || [[ x"$SERVICE_WARM_START" == x"true" ]]; then
+        WARM_BOOT="true"
+    else
+        WARM_BOOT="false"
+    fi
+}
+
+function validate_restore_count()
+{
+    if [[ x"$WARM_BOOT" == x"true" ]]; then
+        RESTORE_COUNT=`$SONIC_DB_CLI STATE_DB hget "WARM_RESTART_TABLE|teamd" restore_count`
+        # We have to make sure db data has not been flushed.
+        if [[ -z "$RESTORE_COUNT" ]]; then
+            WARM_BOOT="false"
+        fi
+    fi
+}
+
+function check_fast_boot ()
+{
+    if [[ $($SONIC_DB_CLI STATE_DB GET "FAST_REBOOT|system") == "1" ]]; then
+        FAST_BOOT="true"
+    else
+        FAST_BOOT="false"
+    fi
+}
+
+start() {
+    debug "Starting ${SERVICE}$DEV service..."
+
+    check_warm_boot
+    validate_restore_count
+
+    check_fast_boot
+
+    debug "Warm boot flag: ${SERVICE}$DEV ${WARM_BOOT}."
+    debug "Fast boot flag: ${SERVICE}$DEV ${Fast_BOOT}."
+
+    # start service docker
+    /usr/bin/${SERVICE}.sh start $DEV
+    debug "Started ${SERVICE}$DEV service..."
+
+}
+
+wait() {
+    /usr/bin/${SERVICE}.sh wait $DEV
+}
+
+stop() {
+    debug "Stopping ${SERVICE}$DEV service..."
+
+    check_warm_boot
+    check_fast_boot
+    debug "Warm boot flag: ${SERVICE}$DEV ${WARM_BOOT}."
+    debug "Fast boot flag: ${SERVICE}$DEV ${FAST_BOOT}."
+
+    if [[ x"$WARM_BOOT" == x"true" ]]; then
+        # Send USR1 signal to all teamd instances to stop them
+        # It will prepare teamd for warm-reboot
+        # Note: We must send USR1 signal before syncd, because it will send the last packet through CPU port
+        docker exec -i ${SERVICE}$DEV pkill -USR1 ${SERVICE} > /dev/null || [ $? == 1 ]
+		
+    elif [[ x"$FAST_BOOT" == x"true" ]]; then
+        # Kill teamd processes inside of teamd container with SIGUSR2 to allow them to send last LACP frames
+        # We call `docker kill teamd` to ensure the container stops as quickly as possible,
+        # then immediately call `systemctl stop teamd` to prevent the service from
+        # restarting the container automatically.
+        # Note: teamd must be killed before syncd, because it will send the last packet through CPU port
+        docker exec -i ${SERVICE}$DEV pkill -USR2 ${SERVICE} || [ $? == 1 ]
+        while docker exec -i ${SERVICE}$DEV pgrep ${SERVICE} > /dev/null; do
+            sleep 0.05
+        done
+        docker kill ${SERVICE}$DEV  &> /dev/null || debug "Docker ${SERVICE}$DEV is not running ($?) ..."
+        systemctl stop ${SERVICE}$DEV
```

Comment:
Let's remove this systemctl stop, it feels like a circular dependency.

Reply:
Looks like this line was added in https://github.com/Azure/sonic-utilities/pull/572. We don't need it any more?

Reply:
As we discussed offline. This line could cause recursive service stop.